scipy-stubs/signal/windows/_windows.pyi

```diff
@@ -29,36 +34,148 @@ __all__ = [
+
+_Window: TypeAlias = Literal[
+    "barthann",
+    "brthan",
+    "bth",
+    "bartlett",
+    "bart",
+    "brt",
+    "blackman",
+    "black",
+    "blk",
+    "blackmanharris",
+    "blackharr",
+    "bkh",
+    "bohman",
+    "bman",
+    "bmn",
+    "boxcar",
+    "box",
+    "ones",
+    "rect",
+    "rectangular",
+    "cosine",
+    "halfcosine",
+    "exponential",
+    "poisson",
+    "flattop",
+    "flat",
+    "flt",
+    "hamming",
+    "hamm",
+    "ham",
+    "hann",
+    "han",
+    "lanczos",
+    "sinc",
+    "nuttall",
+    "nutl",
+    "nut",
+    "parzen",
+    "parz",
+    "par",
+    "taylor",
+    "taylorwin",
+    "triangle",
+    "triang",
+    "tri",
+    "tukey",
+    "tuk",
+]
+_WindowNeedsParams: TypeAlias = Literal[
+    "chebwin",
+    "cheb",
+    "dpss",
+    "gaussian",
+    "gauss",
+    "gss",
+    "general cosine",
+    "general_cosine",
+    "general gaussian",
+    "general_gaussian",
+    "general gauss",
+    "general_gauss",
+    "ggs",
+    "general hamming",
+    "general_hamming",
+    "kaiser",
+    "ksr",
+    "kaiser bessel derived",
+    "kbd",
+]
+
+def general_cosine(M: int, a: Sequence[AnyReal], sym: bool = True) -> _Array_f8_1d: ...
+def boxcar(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def triang(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def parzen(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def bohman(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def blackman(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def nuttall(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def blackmanharris(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def flattop(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def bartlett(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def hann(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def tukey(M: int, alpha: AnyReal = 0.5, sym: bool = True) -> _Array_f8_1d: ...
+def barthann(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def general_hamming(M: int, alpha: AnyReal, sym: bool = True) -> _Array_f8_1d: ...
+def hamming(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def kaiser(M: int, beta: AnyReal, sym: bool = True) -> _Array_f8_1d: ...
+def kaiser_bessel_derived(M: int, beta: AnyReal, *, sym: bool = True) -> _Array_f8_1d: ...
+def gaussian(M: int, std: AnyReal, sym: bool = True) -> _Array_f8_1d: ...
+def general_gaussian(M: int, p: AnyReal, sig: AnyReal, sym: bool = True) -> _Array_f8_1d: ...
+def chebwin(M: int, at: AnyReal, sym: bool = True) -> _Array_f8_1d: ...
+def cosine(M: int, sym: bool = True) -> _Array_f8_1d: ...
+def exponential(M: int, center: AnyReal | None = None, tau: AnyReal = 1.0, sym: bool = True) -> _Array_f8_1d: ...
+def taylor(M: int, nbar: int = 4, sll: int = 30, norm: bool = True, sym: bool = True) -> _Array_f8_1d: ...
+def lanczos(M: int, *, sym: bool = True) -> _Array_f8_1d: ...
+
+#
+@overload
 def dpss(
-    M,
-    NW,
-    Kmax: Untyped | None = None,
+    *,
+    M: int,
```

Comment:
I was considering making a specific int type (private to this module) for this parameter but it seemed like a bit of a pain and `int` seemed like the intention even if it could accept other types.

Reply:
Maybe

```python
_WindowLength: TypeAlias = int | np.int16 | np.int32 | np.int64
```
?

Reply:
> This is indeed an annoying one to annotate:
> 
> After experimenting a bit with it, it turns out that, besides `builtins.int`, it also accepts signed `short`, `intc`, `intp`, `long` and `longlong`, which boils down to `np.int16 | np.int32 | np.int64`. So unsigned integers aren't included. And for some reason, `np.int8` isn't allowed, so `np.signedinteger[Any]` would be too wide.
> 
> I'm guessing that this is also the case for the `M` param of the other functions in this module

This seems to expose a mistyping in the `linalg` module. The reason why `np.int8` doesn't work is because `dpss` calls `linalg.eigh_tridiagonal` using `M` as part of the `select_range` argument which is currently typed in `scipy-stubs` as

```python
_EigSelectRange: TypeAlias = tuple[spt.AnyReal, spt.AnyReal]
```

which allows `np.int8` even though it seems to crash.

Reply:
Looking at the crashing function, it checks that the dtype is one of these `"hilqp"`, which are the types you listed on top, but only if `select="i"`.

Reply:
Good catch! I opened https://github.com/jorenham/scipy-stubs/issues/154 for this.

Reply:
Fixed in https://github.com/jorenham/scipy-stubs/pull/165